From 619acecabf22facd7f523af1c8e97fb9fb4cb74f Mon Sep 17 00:00:00 2001
From: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Date: Fri, 21 Oct 2022 10:12:06 +0300
Subject: [PATCH 19/49] s32cc: Fix system resume when running with SCP

This skips VR5510 initialization as this is in SCP's
ownership and sets A53's reset address during resume.

Issue: ALB-9428
Upstream-Status: Pending 

Signed-off-by: Ghennadi Procopciuc <ghennadi.procopciuc@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 plat/nxp/s32/s32g/s32g_bl2_el3.c |  3 +++
 plat/nxp/s32/s32g/s32g_resume.c  | 13 +++++++++++--
 2 files changed, 14 insertions(+), 2 deletions(-)

diff --git a/plat/nxp/s32/s32g/s32g_bl2_el3.c b/plat/nxp/s32/s32g/s32g_bl2_el3.c
index 725a6ab01..bb209705d 100644
--- a/plat/nxp/s32/s32g/s32g_bl2_el3.c
+++ b/plat/nxp/s32/s32g/s32g_bl2_el3.c
@@ -68,6 +68,9 @@ static int init_and_setup_pmic(void)
 {
 	int ret = 0;
 
+	if (is_scp_used())
+		return 0;
+
 	dt_init_ocotp();
 	dt_init_pmic();
 
diff --git a/plat/nxp/s32/s32g/s32g_resume.c b/plat/nxp/s32/s32g/s32g_resume.c
index 41f1d1849..d9242d483 100644
--- a/plat/nxp/s32/s32g/s32g_resume.c
+++ b/plat/nxp/s32/s32g/s32g_resume.c
@@ -3,17 +3,26 @@
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
-#include <s32gen1-wkpu.h>
-#include <s32_linflexuart.h>
 #include <bl31/bl31.h>		/* for bl31_warm_entrypoint() */
+#include <s32_bl_common.h>
+#include <s32_linflexuart.h>
+#include <s32_lowlevel.h>
+#include <s32gen1-wkpu.h>
 
 void s32g_resume_entrypoint(void)
 {
+	uintptr_t core_addr;
+
 	s32gen1_wkpu_reset();
 
 #if (S32_USE_LINFLEX_IN_BL31 == 1)
 	console_s32_register();
 #endif
 
+	if (is_scp_used()) {
+		core_addr = (uintptr_t)plat_secondary_cold_boot_setup;
+		scp_set_core_reset_addr(core_addr);
+	}
+
 	bl31_warm_entrypoint();
 }
-- 
2.25.1

